L1Trigger/L1TTrackMatch/plugins/L1TrackJetClustering.h

```diff
@@ -113,6 +113,76 @@ namespace l1ttrackjet {
     return PassQuality;
   }
 
+  // Eta binning following the hardware logic
+  inline unsigned int eta_bin_firmwareStyle(int eta_word) {
+    //Function that reads in 15-bit eta word (in two's complement) and returns the index of eta bin in which it belongs
+    //Logic follows exactly according to the firmware
+    //We will first determine if eta is pos/neg from the first bit. Each half of the grid is then split into four coarse bins. Bits 2&3 determine which coarse bin to assign
+    //The coarse bins are split into 5 fine bins, final 13 bits determine which of these coarse bins this track needs to assign
+
+    int eta_coarse = 0;
+    int eta_fine = 0;
+
+    if (eta_word & (1 << 15)) {  //If eta is negative (first/leftmost bit is 1)
+      //Second and third bits contain information about which of four coarse bins
+      eta_coarse = 5 * ((eta_word & (3 << 13)) >> 13);
+    } else {  //else eta is positive (first/leftmost bit is 0)
+      eta_coarse = 5 * (4 + ((eta_word & (3 << 13)) >> 13));
+    }
+
+    //Now get the fine bin index. The numbers correspond to the decimal representation of fine bin edges in binary
+    int j = eta_word & 8191;
+    if (j < 1638)
+      eta_fine = 0;
+    else if (j < 3276)
+      eta_fine = 1;
+    else if (j < 4914)
+      eta_fine = 2;
+    else if (j < 6552)
+      eta_fine = 3;
+    else
+      eta_fine = 4;
+
+    //Final eta bin is coarse bin index + fine bin index, subtract 8 to make eta_bin at eta=-2.4 have index=0
```

Comment:
Stylistically, I wonder if the coarse bin part of this implementation could mirror FastHisto's usage of adding half the track z0 max and truncating the result to get a bin index for histogramming (requiring an integer N:M mapping between histogram bins and trackword information) . Probably not worth implementing right now, though.

Reply:
We might be able to look into this later if deemed necessary